### PR TITLE
Add initial version of the style engine

### PR DESCRIPTION
**Pull request by youknowriad**
This is an exploration to try to find a path towards what's being discussed in https://github.com/WordPress/gutenberg/discussions/37495

My main goal here is not to refactor everything but to try to bootstrap a way for us to iterate without breaking changes or anything and build that engine iteratively until we can pull the plug entirely. 

This is very early but worth sharing to start concrete discussions

## Features include

* Added a new `packages/style-engine` package.
* JS implementation of the style engine exposes a `generate` function that takes a style object and an options object (which allows for defining a selector), and returns rendered CSS.
* For the shorter-term, a `getCSSRules` function is also exposed, that returns the rendered styles in an array of objects that provides a `selector`, a `key` that matches the CSS property used, and a `value`.
* Documentation for the package also includes a to-do-list for further things to work on in follow-ups

## How to test this change

* This PR seeks to only minimally affect how things work. Test that adding and changing padding of an individual block (e.g. Group) works correctly in the post and site editors. Also test adjusting padding within global styles. The scope of the proposed change is just within the editor, not within the server-rendered view of the front end of the site.

### CODE REVIEW

**Comment by andrewserong** [comment]
Nice work, thanks for getting this started @youknowriad! Lots of great decisions here, I like the simplicity of the API that gets exposed (`generate` and `getCSSrules`). I imagine the PHP implementation will have a corresponding API so that it's roughly the same process on the frontend / backend? I see a couple of comments that the goal is to move everything to server side generated engine styles, which sounds good for removing duplication long-term.

This might be premature, but one thing I'm curious about, for getting parity between frontend and backend styles, and being able to one-day resolve #35376 is how we might go about getting / giving access to the `style` object in the context of previews within the post content block? (Or to put it more generally: getting styles for post content retrieved in isolation via the REST API)

File: packages/style-engine/src/types.ts
Comment (andrewserong) [inline packages/style-engine/src/types.ts]: Tiny nit: should this attribute and all the ones beneath it be in an optional `typography?` object instead of `spacing?`

**Review by aaronrobertshaw — COMMENTED** [review]
Amazing start so far @youknowriad 🙇 

> Lots of great decisions here, I like the simplicity of the API that gets exposed (generate and getCSSrules).

Exactly. The API looks great and gets another +1 from me.

I think one issue with the current block supports, global styles and theme.json has been that the documentation has been a bit lacking in terms of providing devs with easily consumed real-world examples. In many cases, it relies on people reading through the code to fully understand how it all comes together before they can then apply that knowledge to their use case.

Given the discussions to date around the styles engine effectively give us a checklist of features we'd like it to have. Would it be prudent to add usage examples as these features and capabilities get added? I appreciate it's very early days but could save a lot of "support" in the future.

File: packages/style-engine/src/types.ts
Comment (aaronrobertshaw) [reply packages/style-engine/src/types.ts]: I spotted this as well. I believe you are correct and it will be ironed out as this PR evolves.

File: packages/block-editor/src/hooks/style.js
Comment (ramonjd) [inline packages/block-editor/src/hooks/style.js]: I think this should be `if ( ! STYLE_PROPERTY[ propKey ].useEngine ) {` ?

File: packages/block-editor/src/hooks/style.js
Comment (ramonjd) [inline packages/block-editor/src/hooks/style.js]: Will `getCSSRules` need to return React-safe style camel case props as well, e.g., `{ style: { paddingTop: '19px', ... } }`

I'm seeing mixed output:

Screen Shot 2022-01-17 at 12 41 15 pm

I know the focus is on rendering styles to the page, so apologies if I'm jumping the 🔫

**Comment by ramonjd** [comment]
Thanks for starting this and giving us a sneak preview of your idea ❤️ 

Very curious to see how this develops, in particular, how/whether could ultimately leverage a backend implementation to register, dedupe and bundle CSS definitions in one `<style />` block. 

Screen Shot 2022-01-17 at 12 55 22 pm

Probably a bit down the track!

File: packages/block-editor/src/hooks/style.js
Comment (youknowriad) [reply packages/block-editor/src/hooks/style.js]: Indeed, I guess I didn't pay enough attention here :).

For me `getCSSRules` was supposed to be an internal API (not exposed) but I had to find something to make the current style hook and theme.json style renderer continue to work with the padding moved to the style engine. So I just used what I had. I'll need to check where the mixing happens and just pick one for now.

**Comment by andrewserong** [comment]
I'm just adding myself as an assignee. For anyone else reading on, I've offered to take over this PR from Riad, so will give it a rebase and work through some of the things we encountered when we first took a look at.

File: packages/block-editor/src/hooks/style.js
Comment (andrewserong) [reply packages/block-editor/src/hooks/style.js]: Updated in https://github.com/WordPress/gutenberg/pull/37978/commits/e1e231db3a2efea5ba4ebf80ffb41a3d4b68f22b

File: packages/style-engine/src/types.ts
Comment (andrewserong) [reply packages/style-engine/src/types.ts]: I've moved this out to a separate `typography` object in https://github.com/WordPress/gutenberg/pull/37978/commits/24f8cb37504fe94efa0c9e094ceede3b8529cb00

File: packages/blocks/src/api/constants.js
Comment (youknowriad) [inline packages/blocks/src/api/constants.js:103]: Ideally at the end of the style engine implementation, this "constant" might be removed entirely if we did our work properly :)

File: packages/edit-site/src/components/global-styles/use-global-styles-output.js
Comment (youknowriad) [inline packages/edit-site/src/components/global-styles/use-global-styles-output.js]: I think we should probably ignore the properties with useEngine: true here kind of like https://github.com/WordPress/gutenberg/pull/37978/files#diff-292053c9ec98ebdd73943767587e8093315ad3a2bf01ca3b5352efc48326becdR89

File: packages/block-editor/src/hooks/style.js
Comment (andrewserong) [reply packages/block-editor/src/hooks/style.js]: Interesting problem here. I had a play with a couple of different approaches and settled on switching `getCSSRules` to return the camelCase keys (since it's an array of JS objects, it felt slightly more natural to me for it to be using the JS style attribute convention), and then get `generate` to be responsible for converting it to kebab-case when outputting the CSS styles. Updated in this commit: https://github.com/WordPress/gutenberg/pull/37978/commits/a756e443a95da01dc20493099a96668ee8b20ab9

That seemed to fix it for me, though it does introduce importing another couple of `lodash` functions. I imagine we might want to see if we can get rid of the `lodash` dependency if we're expecting the style engine to also be run in standalone frontend environments outside of Gutenberg, but for the moment at least, it made it quicker to get the fix in.

Happy to change the behaviour here if folks have a preference for the shape of the attributes!

File: packages/edit-site/src/components/global-styles/use-global-styles-output.js
Comment (andrewserong) [reply packages/edit-site/src/components/global-styles/use-global-styles-output.js]: Good catch! I've added `useEngine` to the check that bails early in the loop in https://github.com/WordPress/gutenberg/pull/37978/commits/31981a16172a30f8d49d47b55f5a889c8121adc6 — I've also updated the output slightly to match the change to `getCSSRules` [discussed here](https://github.com/WordPress/gutenberg/pull/37978#discussion_r791393936).

**Comment by andrewserong** [comment]
I've made a couple of updates to the PR and the unit tests are passing now, with the change in `getCSSRules` to use camelCase for the CSS properties. I'll be AFK tomorrow, but let me know if there are any other changes folks would like to see at this stage, and I will pick this up again when I'm back on Thursday!

**Comment by andrewserong** [comment]
I've updated the PR description with a little more detail as to the scope of this PR, and switched it over to ready for review to gather more feedback.

What do folks think about what's included currently in this PR? At the initial stage, do we need to add anything else just yet?

For example, we'll need to think about how we'd like the server-rendered implementation to work — is that something to consider at this stage in this PR, or something to explore in a follow-up? I imagine the `php` implementation will live in `/lib` rather than within the `style-engine` package, so in terms of code-organisation, it might be kept a little separate.

@youknowriad — did you have anything in particular in mind for the PHP-side, or is that something for us to explore separately?

**Comment by youknowriad** [comment]
> @youknowriad — did you have anything in particular in mind for the PHP-side, or is that something for us to explore separately?

I was still hesitant about whether to include the php side in this PR or as follow-up, I guess both can work.

In terms of code though, I still think it might be good to keep it in the package itself (kind of like we have php code in block-library). I'm sure it has some challenges in terms of "building the plugins", "auto-renaming functions for the Gutenberg plugin", and "auto-merging code into Core" but the same issues are "solved" for the block-library php code as well.

**Comment by andrewserong** [comment]
> In terms of code though, I still think it might be good to keep it in the package itself 

Thanks @youknowriad, I love that idea, I think it'd be much neater to keep it all together. Since that adds a little more complexity to setting it up, I think it'd be worth pursuing that in a follow-up after this PR lands since we're already at 21 changed files, and it might make it easier to tease apart the PHP side of things in isolation.

File: packages/style-engine/src/index.ts
Comment (andrewserong) [inline packages/style-engine/src/index.ts]: Just thinking, should this signature be a `style` object and an `options` object, instead of the selector being the second param? Then we'd have `selector` as one of the options in the `options` object.

If so, we might do the same for `getCSSRules`.

Comment (ramonjd) [reply packages/style-engine/src/index.ts]: Sounds legit 😆 

Are there further options that we need in these methods, or is it about setting up API for flexibility?

Comment (youknowriad) [reply packages/style-engine/src/index.ts]: I think `options` could make sense here yeah, things like "enabled styles" for instance could be an option, or a flag on whether to optimize or not... 

That said, I'm not too concerned because that package is written as "private" allowing us to make breaking changes before actually considering making it "public".

Comment (andrewserong) [reply packages/style-engine/src/index.ts]: Thanks for the feedback! I thought I'd have a go at it now, so that we don't have to refactor it separately when we go to add another option. I've added an `options` object in https://github.com/WordPress/gutenberg/pull/37978/commits/a925476e0c3d492ec6e6b974949c7bf85be215a7, added an additional type, and pass the object around to the style definitions.

> Are there further options that we need in these methods, or is it about setting up API for flexibility?

Yes, it's about flexibility — we don't need to add additional options _just_ yet for this PR, but we do know that we'd like to explore adding additional ones, so 🤞 switching to `options` now makes it a bit easier for exploratory follow-up PRs.

Let me know if anyone would like the implementation changed, though, of course!

File: packages/style-engine/src/styles/padding.ts
Comment (ramonjd) [inline packages/style-engine/src/styles/padding.ts:13]: Could this path value one day be an options value we get from https://github.com/WordPress/gutenberg/blob/4452f092259b93521440aa0a8b01b872cc1350de/packages/blocks/src/api/constants.js#L94?

Just wondering if the style engine needs to know about the block implementation detail, or if we even need a single source of ~vermouth~ truth.

File: packages/block-editor/src/hooks/style.js
Comment (ramonjd) [inline packages/block-editor/src/hooks/style.js]: Could this be `output[ rule.key ] = rule.value`? Seems there's no reason to exit the callback scope with `return` when there's nothing after. Or have I missed something?

**Comment by ramonjd** [comment]
I just gave this another walkthrough and it's working invisibly (that is, great!) for padding values in the block and site editors. 

I imagine things will start to coalesce once we start supporting more style properties, and looking at layout and so on.

Thanks a lot for running with this @andrewserong !

File: packages/block-editor/src/hooks/style.js
Comment (andrewserong) [reply packages/block-editor/src/hooks/style.js]: Oh, good point. I think it might have been leftover from a previous way of doing it. I'll update 👍

File: packages/style-engine/src/styles/padding.ts
Comment (andrewserong) [reply packages/style-engine/src/styles/padding.ts:13]: Good question! I think having a single source of truth is a good idea, but I imagine that might actually live in this package instead of in the blocks package (if we assume that the style engine is the "owner" of how to process the style object)? I suppose we could either move the constants over to the style engine in a central place, or leave it up to each individual style definition to know only about itself?

Either way, probably a good thing for us to explore in follow-ups!

**Comment by andrewserong** [comment]
@WordPress/gutenberg-core — just a heads-up that I think I've updated this PR to the point where it could use more feedback / testing or input from more core folks 😀

I think at this stage, one of the main questions is whether or not the style engine as implemented in this PR is a good place for us to merge and then explore iterating on in subsequent PRs (including adding in server-side behaviour). Before this PR lands, though, I wanted to make sure we get enough eyes on the direction it's currently taking, so that we can confidence check the approach.

There is no rush on this at all, though — ideally the work here should be iterative and not-blocking on other work underway 🙂

**Comment by Luehrsen** [comment]
If I understand this PR correctly the style engine defines an own, custom standard to define, maintain and manipulate style rules and declarations.

Maybe it's worth looking at what other implementations are doing. postCSS and connected projects are using a concept called "[Abstract Syntax Trees](https://css-tricks.com/how-to-modify-nodes-in-an-abstract-syntax-tree/)" to hold and manipulate the parsed CSS. 

Maybe it's even worth exploring if we could store the global stylesheet as a JSON representation of the AST. The advantage would be a language agnostic representation of the current state of the global style, that can be filtered, rendered and modified either serverside or clientside.

- https://github.com/csstree/csstree
- https://postcss.org/api/
- https://astexplorer.net/

Thank you for tackling this issue! Having a lean and customizable stylsheet is really important for us.

**Comment by andrewserong** [comment]
Thanks for taking a look @Luehrsen and for all the great links, that gives us lots to read up on! That's a really interesting idea, and it looks like ASTs are currently in use for the [gradient color picker](https://github.com/WordPress/gutenberg/blob/8e04929799f21ff6ad07f60dfe5fa555c7edfcdd/packages/components/src/custom-gradient-picker/utils.js#L32) in Gutenberg (via `gradient-parser`), so I'm sure that'll be a really valuable direction to explore.

I suppose the goal at this stage, is to see if we can come up with a minimal (and viable) enough idea for how we call and return values / styles via the style engine, so that in subsequent PRs we can explore how it'll all work internally, in a way where we can replace things / make changes as we go.

**Comment by youknowriad** [comment]
Let's get this in and iterate. I think it's a good start.

**Comment by andrewserong** [comment]
Thanks @youknowriad! It sounds like there was interest from one or two folks to give this PR a look over the weekend, so I'm happy to wait until Monday to give it a final rebase and merge it in, if there are no objections.

**Comment by luisherranz** [comment]
> It sounds like there was interest from one or two folks to give this PR a look

I don't know if you are referring to me, I guess not. But as we talked about the implications of this PR for frontend hydration, I'll add my review here:

**It looks great to me!! 👏** 

That's it 😄

File: packages/style-engine/src/index.ts
Comment (luisherranz) [inline packages/style-engine/src/index.ts]: Minor comment: Why do we need immutability here?

File: packages/style-engine/src/types.ts
Comment (ciampo) [inline packages/style-engine/src/types.ts:6]: Have you considered using the `React.CSSProperties` types for most types in this file?

e.g

```ts

type Props = {
  backgroundColor: CSSProperties[ 'backgroundColor' ];
}
```

Comment (andrewserong) [reply packages/style-engine/src/types.ts:6]: Good question, we can probably move most of them over to `CSSProperties` (e.g. `fontWeight`, `fontStyle`, etc). For the `Box` type, since it's used in both margin and padding, it might need to be a bit more flexible than targeting a single CSS property. What do you think would work for that type?

File: packages/style-engine/src/index.ts
Comment (andrewserong) [reply packages/style-engine/src/index.ts]: I don't think it's so much about immutability, but concatenating the array returned from `definition.generate()`. So the following might be roughly equivalent to what we want to do here?

```
rules = rules.concat( definition.generate( style, options ) );
```

If we have a lot of rules, it's potentially a bit faster using `concat` than the spread syntax, too, so happy to change it over.

File: packages/style-engine/src/types.ts
Comment (ciampo) [reply packages/style-engine/src/types.ts:6]: Mhh, not sure about that. The main thing is that the idea of `Box` (an object with `top` `right` `bottom` and `left` props) doesn't really translate to CSS.

If we need to maintain the structure of these types, we could still use `CSSProperties` where we're using `string`, and leave the rest unchanged:

```ts
export type Box = {
	top?: CSSProperties[ 'top' ];
	right?: CSSProperties[ 'right' ];
	bottom?: CSSProperties[ 'bottom' ];
	left?: CSSProperties[ 'left' ];
};

export interface Style {
	spacing?: {
		margin?: CSSProperties[ 'margin' ] | Box;
		padding?: CSSProperties[ 'padding' ] | Box;
	};
	// ...
};
```

WDYT?

File: packages/style-engine/src/index.ts
Comment (luisherranz) [reply packages/style-engine/src/index.ts]: Why not simply add the new rules to the array?

```js
rules.push( ...definition.generate( style, options ) );
```

Comment (andrewserong) [reply packages/style-engine/src/index.ts]: That works, too! 😄
I've pushed an update in https://github.com/WordPress/gutenberg/pull/37978/commits/91109e1a9c4231014f3a1368c39b5c61bcf9fffa

File: packages/style-engine/src/types.ts
Comment (andrewserong) [reply packages/style-engine/src/types.ts:6]: Oh, I quite like that, and it lets us keep the `Box` abstraction. It isn't _quite_ right because it's referring to different properties, but the underlying type should be much the same, and it still serves the purpose of better communicating the desired type. I've updated in https://github.com/WordPress/gutenberg/pull/37978/commits/f0cbeffdc5c0854fa48152e368f1fcfe92871156, and we can continue to tweak the types in follow-ups, too 🙂

**Comment by andrewserong** [comment]
I've implemented the latest couple of bits of feedback and given this a rebase. It's still technically the weekend in other parts of the world, so I'll leave this overnight my time and merge in tomorrow if there are no objections!

File: packages/style-engine/src/types.ts
Comment (ciampo) [reply packages/style-engine/src/types.ts:6]: Thank you! I think it's already much better.

> It isn't quite right because it's referring to different properties, but the underlying type should be much the same,

Out of curiosity, what did you mean with "referring to different properties" ?

Comment (andrewserong) [reply packages/style-engine/src/types.ts:6]: > what did you mean with "referring to different properties" ?

Apologies, I could have been clearer there! What I meant is that `CSSProperties[ 'top' ]` refers to the [top](https://developer.mozilla.org/en-US/docs/Web/CSS/top) property and not for example [padding-top](https://developer.mozilla.org/en-US/docs/Web/CSS/padding-top) — but since they both accept pretty much the same actual values, it doesn't matter too much that the type isn't precise. I think `CSSProperty[ 'top' ]` is preferable to a more complex type like `CSSProperty[ 'top' ] | CSSProperty[ 'paddingTop ' ] | CSSProperty[ 'marginTop' ]` which could get a bit unwieldy, even if it's technically more accurate.

But, happy to explore in follow-ups, of course!

**Review by andrewserong — APPROVED** [review]
I think this is good to land now, and there've been no objections in proceeding further, so I'll give it the ✅😀

Thanks again for all the feedback and testing, folks! Let's continue discussions on the tracking issue (https://github.com/WordPress/gutenberg/issues/38167) and related discussions.

File: packages/style-engine/src/types.ts
Comment (ciampo) [reply packages/style-engine/src/types.ts:6]: Proposed a potential follow-up in #38894

**Pull request merged**
